agent/taskresource/volume/dockervolume.go

```diff
@@ -71,40 +91,124 @@ func NewVolumeResource(name string,
+	if scope == TaskScope {
+		resourceStatusToTransitionFunction[taskresource.ResourceStatus(VolumeRemoved)] = vol.Cleanup
+	}
+
+	vol.resourceStatusToTransitionFunction = resourceStatusToTransitionFunction
+}
+
+// GetName returns the name of the volume resource
+func (vol *VolumeResource) GetName() string {
+	return vol.Name
+}
+
+// DesiredTerminal returns true if the cgroup's desired status is REMOVED
+func (vol *VolumeResource) DesiredTerminal() bool {
+	vol.lock.RLock()
+	defer vol.lock.RUnlock()
+
+	return vol.desiredStatusUnsafe == taskresource.ResourceStatus(VolumeRemoved)
 }
 
 // SetDesiredStatus safely sets the desired status of the resource
-func (vol *VolumeResource) SetDesiredStatus(status VolumeStatus) {
+func (vol *VolumeResource) SetDesiredStatus(status taskresource.ResourceStatus) {
 	vol.lock.Lock()
 	defer vol.lock.Unlock()
 
 	vol.desiredStatusUnsafe = status
 }
 
 // GetDesiredStatus safely returns the desired status of the task
-func (vol *VolumeResource) GetDesiredStatus() VolumeStatus {
+func (vol *VolumeResource) GetDesiredStatus() taskresource.ResourceStatus {
 	vol.lock.RLock()
 	defer vol.lock.RUnlock()
 
 	return vol.desiredStatusUnsafe
 }
 
 // SetKnownStatus safely sets the currently known status of the resource
-func (vol *VolumeResource) SetKnownStatus(status VolumeStatus) {
+func (vol *VolumeResource) SetKnownStatus(status taskresource.ResourceStatus) {
 	vol.lock.Lock()
 	defer vol.lock.Unlock()
 
 	vol.knownStatusUnsafe = status
 }
 
 // GetKnownStatus safely returns the currently known status of the task
-func (vol *VolumeResource) GetKnownStatus() VolumeStatus {
+func (vol *VolumeResource) GetKnownStatus() taskresource.ResourceStatus {
 	vol.lock.RLock()
 	defer vol.lock.RUnlock()
 
 	return vol.knownStatusUnsafe
 }
 
+// KnownCreated returns true if the volume's known status is CREATED
+func (vol *VolumeResource) KnownCreated() bool {
+	vol.lock.RLock()
+	defer vol.lock.RUnlock()
+
+	return vol.knownStatusUnsafe == taskresource.ResourceStatus(VolumeCreated)
+}
+
+// TerminalStatus returns the last transition state of volume
+func (vol *VolumeResource) TerminalStatus() taskresource.ResourceStatus {
+	return taskresource.ResourceStatus(VolumeRemoved)
+}
+
+// NextKnownState returns the state that the resource should
+// progress to based on its `KnownState`.
+func (vol *VolumeResource) NextKnownState() taskresource.ResourceStatus {
+	return vol.GetKnownStatus() + 1
+}
+
+// SteadyState returns the transition state of the resource defined as "ready"
+func (vol *VolumeResource) SteadyState() taskresource.ResourceStatus {
+	return taskresource.ResourceStatus(VolumeCreated)
+}
+
+// ApplyTransition calls the function required to move to the specified status
+func (vol *VolumeResource) ApplyTransition(nextState taskresource.ResourceStatus) error {
+	transitionFunc, ok := vol.resourceStatusToTransitionFunction[nextState]
+	if !ok {
+		seelog.Errorf("Volume Resource [%s]: unsupported desired state transition: %s",
```

Comment:
Is this the right level of detail for failing to transition the state of a critical task resource?

Reply:
How would failures here be debugged if, for example, docker was having an issue with the volume that bubbled up here and resulted in a timeout for a state transition?